### PR TITLE
[Merged by Bors] - feat(NumberTheory/LSeries): analytic continuation of Dirichlet L-series

### PR DESCRIPTION
**Pull request by loefflerd**
We prove analytic continuation of the L-function of any periodic function on the naturals, including Dirichlet characters.

---

- [x] depends on: #13353 
- [x] depends on: #14495

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/LSeries/ZMod.lean
Comment (jcommelin) [inline Mathlib/NumberTheory/LSeries/ZMod.lean:93]: Could you please consider writing this as a `calc`-mode proof? I think it would improve readability, and probably also maintainability of the proof.

File: Mathlib/NumberTheory/LSeries/DirichletContinuation.lean
Comment (jcommelin) [inline Mathlib/NumberTheory/LSeries/DirichletContinuation.lean:44]: Should this be a simp lemma?

File: Mathlib/NumberTheory/LSeries/ZMod.lean
Comment (jcommelin) [inline Mathlib/NumberTheory/LSeries/ZMod.lean:159]: Minor suggestion:
```suggestion
  have hU {u} : u ∈ U ↔ u ≠ 1 ∨ j ≠ 0 := by simp only [Set.mem_ite_univ_right, U]; tauto
  -- hypotheses for uniqueness of analytic continuation
  have hf : AnalyticOn ℂ f U := by
    refine DifferentiableOn.analyticOn (fun u hu ↦ ?_) hUo
    refine (differentiableAt_LFunction _ _ ((hU.mp hu).imp_right fun h ↦ ?_)).differentiableWithinAt
    simp only [mul_comm j, AddChar.sum_mulShift _ (isPrimitive_stdAddChar _), h,
        ↓reduceIte, CharP.cast_eq_zero, or_true]
  have hg : AnalyticOn ℂ g U := by
    refine DifferentiableOn.analyticOn (fun u hu ↦ ?_) hUo
    refine (differentiableAt_expZeta _ _ ((hU.mp hu).imp_right fun h ↦ ?_)).differentiableWithinAt
    rwa [ne_eq, toAddCircle_eq_zero]
  have hUc : IsPreconnected U := by
    by_cases h : j = 0
    · simpa only [h, ↓reduceIte, U] using
        (isConnected_compl_singleton_of_one_lt_rank (by simp) _).isPreconnected
    · simpa only [h, ↓reduceIte, U] using isPreconnected_univ
  have hV : V ∈ 𝓝 2 := (continuous_re.isOpen_preimage _ isOpen_Ioi).mem_nhds (by simp)
  have hUmem : 2 ∈ U := by simp [U]
  have hUmem' : s ∈ U := hU.mpr hjs.symm
```

File: Mathlib/Analysis/SumOverResidueClass.lean
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/SumOverResidueClass.lean:99]: This seems like a purely algebraic result; can it go somewhere earlier?

File: Mathlib/NumberTheory/LSeries/Basic.lean
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/LSeries/Basic.lean:80]: The assumption in the lemma is `s ≠ 0`, not `0 < re s`

File: Mathlib/NumberTheory/LSeries/DirichletContinuation.lean
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/LSeries/DirichletContinuation.lean:44]: Should we not be using `lfunction` in lemma names by the naming convention? (Or namespacing)

File: Mathlib/NumberTheory/LSeries/ZMod.lean
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/LSeries/ZMod.lean:66]: Why do you need the `change`? Please try to avoid it

File: Mathlib/NumberTheory/LSeries/ZMod.lean
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/LSeries/ZMod.lean:77]: I would avoid relying on the `Set.Icc` definition, maybe with:
```suggestion
  have ha : (j.val / N : ℝ) ∈ Set.Icc 0 1 := by
    rw [Set.mem_Icc]
    constructor
    · positivity
    · rw [div_le_one (Nat.cast_pos.mpr <| NeZero.pos _), Nat.cast_le]
      exact (val_lt _).le
```

File: Mathlib/NumberTheory/LSeries/ZMod.lean
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/LSeries/ZMod.lean:97]:
```suggestion
  Differentiable.const_cpow (by fun_prop) (Or.inl <| NeZero.ne _)
```

File: Mathlib/NumberTheory/LSeries/ZMod.lean
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/LSeries/ZMod.lean:96]: `Npow` doesn't seem in line with the naming convention either, though I guess you get a little more leeway with private lemmas

File: Mathlib/NumberTheory/LSeries/DirichletContinuation.lean
Comment (loefflerd) [reply Mathlib/NumberTheory/LSeries/DirichletContinuation.lean:44]: This is a counterpart of the existing `DirichletCharacter.LSeries_modOne_eq`.

File: Mathlib/Analysis/SumOverResidueClass.lean
Comment (loefflerd) [reply Mathlib/Analysis/SumOverResidueClass.lean:99]: Good point, I moved it to `Data.ZMod.Basic`.

File: Mathlib/NumberTheory/LSeries/ZMod.lean
Comment (loefflerd) [reply Mathlib/NumberTheory/LSeries/ZMod.lean]: It only gets used twice anyway so I inlined it

**Comment by loefflerd** [comment]
Thanks Ruben and Johan for all the comments! Ready for review again.

File: Mathlib/NumberTheory/LSeries/ZMod.lean
Comment (CBirkbeck) [inline Mathlib/NumberTheory/LSeries/ZMod.lean:62]: I'm guessing this example shouldn't be here.

Comment (loefflerd) [reply Mathlib/NumberTheory/LSeries/ZMod.lean]: Oops! Yes, that was just me using `exact?` to find a simp lemma. Fixed now.

File: Mathlib/NumberTheory/LSeries/DirichletContinuation.lean
Comment (CBirkbeck) [inline Mathlib/NumberTheory/LSeries/DirichletContinuation.lean:47]:
```suggestion
    LFunction χ = riemannZeta := by
  ext1; rw [LFunction, ZMod.LFunction_modOne_eq, (by rfl : (0 : ZMod 1) = 1), map_one, one_mul]
```

File: Mathlib/NumberTheory/LSeries/ZMod.lean
Comment (loefflerd) [reply Mathlib/NumberTheory/LSeries/ZMod.lean]: (Off-topic: GitHub's web interface seems to have a bug. Your suggestion was to remove an unnecessary extra blank line, but the web page displays it as if you wanted to _add_ a blank line. Fortunately what actually got committed is the right thing, it is just the preview of the suggested change that is bogus.)

File: Mathlib/NumberTheory/LSeries/ZMod.lean
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/LSeries/ZMod.lean:145]:
```suggestion
      ↓reduceIte, CharP.cast_eq_zero, or_true]
```

**Review by Ruben-VandeVelde — COMMENTED** [review]
I'm not sure I agree with the L-in-names argument, but I see we'd need to change it more widely than this change, so no need to block on it.

Just one more comment from my side

**Comment by loefflerd** [comment]
Well spotted, thanks Ruben! Are we done here, then?

File: Mathlib/NumberTheory/LSeries/DirichletContinuation.lean
Comment (kbuzzard) [inline Mathlib/NumberTheory/LSeries/DirichletContinuation.lean:13]:
```suggestion
L-series of `χ` has analytic continuation (away from a pole at `s = 1` if `χ` is trivial).
```

**Review by kbuzzard — APPROVED** [review]
Excellent work David! Thanks a lot!

**Comment by MichaelStollBayreuth** [comment]
> Excellent work David! Thanks a lot!

+1 (and sorry for not finding the time to review...)

**Comment by loefflerd** [comment]
Thanks for all the feedback, everyone! Is anyone willing to take the leap and submit this for merging?

**Comment by MichaelStollBayreuth** [comment]
OK, I'll bite!
maintainer merge

**Comment by kbuzzard** [comment]
Bors merge

**Comment by kbuzzard** [comment]
bors merge

**Comment by sgouezel** [comment]
bors r-
Can you merge master, fix the lint, and then send it back to bors?
bors d+

**Comment by loefflerd** [comment]
Aargh – it seems that in the interval of ~4 days since I last merged master, some other PR's got merged which added a few lines to `Mathlib.Data.ZMod.Basic`, and those changes together with the ones in this PR put it just over the length limit. Will fix.

File: scripts/style-exceptions.txt
Comment (Parcly-Taxel) [inline scripts/style-exceptions.txt:1]: And then another PR landed, which means you should not add to `style-exceptions.txt` anymore. Instead add `set_option linter.style.longFile 1700` to the bottom of `Data.ZMod.Basic`.

**Review by Parcly-Taxel — CHANGES_REQUESTED** [review]

File: scripts/style-exceptions.txt
Comment (loefflerd) [reply scripts/style-exceptions.txt:1]: > which means you should not add to style-exceptions.txt anymore. 

The linting script said, "run `lake exe lint-style --update` to fix this warning". So I did. Sounds like the code for `lint-style --update` needs updating...

**Pull request closed without merging**
